docs/SIL.rst

```diff
@@ -1579,6 +1590,179 @@ of functions returning uninhabited types. An ``unreachable`` instruction that
+    // without triggering an ownership violation.
+    store %kls2 to [init] %globalMem : $*Klass
+
+    // "Consuming Use" of %kls1.
+    // Definition of %kls1Casted.
+    %kls1Casted = upcast %kls1 : $Klass to $SuperKlass
+
+    // "Consuming Use" of %kls1Casted
+    return %kls1Casted : $SuperKlass
+  }
+
+Notice how every value in the SIL above has a partionable set of uses with
+normal uses always before consuming uses. Any such violations of ownership
+semantics would trigger a static SILVerifier error allowing us to know that we
+do not have any leaks or use-after-frees in the above code.
+
+The semantics in the previous example is of just one form of ownership semantics
+supported: "owned" semantics. In SIL, we support four different ownership kinds:
+
+* **None**. This is used to represent values that do not require memory
+  management like inert "Plain Old Data" values such as an integer or a float.
+
+* **Owned**. A value that exists independently of any other value and is
+  consumed exactly once along all paths through the program by either a
+  destroy_value (actually destroying the value) or by a consuming instruction
+  that rebinds the value (consider an apply or a cast).
+
+* **Guaranteed**. A value with a scoped lifetime whose liveness is dependent on
+  the lifetime of some other "base" owned or guaranteed value. Consumed by
+  end_borrow instructions.
+
+* **Unowned**. A value that is only guaranteed to be instantaneously valid and
+  must be copied before the value is used in an `@owned` or `@guaranteed`
+  context. This is legacy needed to model argument values with the ObjC unsafe
+  unowned argument convention. This value should never be consumed.
+
+We describe each of these semantics in more detail below.
+
+Value Ownership Kind
+~~~~~~~~~~~~~~~~~~~~
+
+Owned
+`````
+
+Owned ownership models "move only" values. We require that each such value is
+consumed exactly once along all program paths. The IR verifier will flag values
+that are not consumed along a path as a leak and any double consumes as
+use-after-frees. We model move operations via "forwarding uses" such as casts
+and transforming terminators (e.x.: switch_enum, checked_cast_br) that transform
+the input value, consuming it in the process, and producing a new transformed
+owned value as a result.
+
+Given any value, one can produce a new "copy" of the owned value via the usage
+of the copy_value instruction. Since in ARC a "copy" of an owned class is just a
+reference count increment, we can use owned values to break down complex ARC
+def-use graphs into individual "move only" values::
+
+  // TODO:
+  // testcase.swift.
+  func doSomething(x : Klass) {
+    let y = x as? OtherKlass
+    switch y {
+    case .none:
+      break
+    case .some(let z):
+      use(z)
+    }
+  }
+
+  // testcase.sil. A possible SILGen lowering
+  sil @doSomething : $@convention(thin) (@guaranteed Klass) -> () {
```

Comment:
I think this SIL example is too large to get the point. I would just pick either the case or the switch for the example.

Reply:
Which do you think is better? I wanted it to be slightly complex to show how the proof builds up as one flows through the SIL. But I see your point.

Reply:
I would pick the cast, but I think both are fine.

Reply:
Ok. I'll fix the cast.